libbeat/processors/rate_limit/algorithm/token_bucket.go

```diff
@@ -50,65 +86,100 @@ func newTokenBucket(config Config) (Algorithm, error) {
 	return &tokenBucket{
 		config.Limit,
 		config.Limit.value * cfg.BurstMultiplier,
-		make(map[string]bucket, 0),
+		sync.Map{},
+		struct {
+			thresholds tokenBucketGCConfig
+			metrics    tokenBucketGCConfig
+		}{
+			thresholds: tokenBucketGCConfig{
+				NumCalls:   cfg.GC.NumCalls,
+				NumBuckets: cfg.GC.NumBuckets,
+			},
+		},
+		clock.RealClock{},
 	}, nil
 }
 
-func (t *tokenBucket) IsAllowed(key string) bool {
-	t.replenishBuckets()
+func (t *tokenBucket) IsAllowed(key uint64) bool {
+	t.runGC()
 
 	b := t.getBucket(key)
 	allowed := b.withdraw()
-	t.buckets[key] = b
 
+	t.gc.metrics.NumCalls++
 	return allowed
 }
 
-func (t *tokenBucket) getBucket(key string) bucket {
-	b, exists := t.buckets[key]
-	if !exists {
-		b = bucket{
-			tokens:        t.depth,
-			lastReplenish: time.Now(),
-		}
-		t.buckets[key] = b
+// SetClock allows test code to inject a fake clock
+func (t *tokenBucket) SetClock(c clock.Clock) {
+	t.clock = c
+}
+
+func (t *tokenBucket) getBucket(key uint64) *bucket {
+	v, exists := t.buckets.LoadOrStore(key, &bucket{
+		tokens:        t.depth,
+		lastReplenish: t.clock.Now(),
+	})
+	b := v.(*bucket)
+
+	if exists {
+		b.replenish(t.limit, t.clock)
+		return b
 	}
 
+	t.gc.metrics.NumBuckets++
 	return b
-
 }
 
 func (b *bucket) withdraw() bool {
-	if b.tokens == 0 {
+	if b.tokens < 1 {
 		return false
 	}
 	b.tokens--
 	return true
 }
 
-func (b *bucket) replenish(rate Rate) {
-	secsSinceLastReplenish := time.Now().Sub(b.lastReplenish).Seconds()
+func (b *bucket) replenish(rate Rate, clock clock.Clock) {
+	secsSinceLastReplenish := clock.Now().Sub(b.lastReplenish).Seconds()
 	tokensToReplenish := secsSinceLastReplenish * rate.valuePerSecond()
 
 	b.tokens += tokensToReplenish
-	b.lastReplenish = time.Now()
+	b.lastReplenish = clock.Now()
 }
 
-func (t *tokenBucket) replenishBuckets() {
-	toDelete := make([]string, 0)
+func (t *tokenBucket) runGC() {
+	// Don't run GC if thresholds haven't been crossed.
+	if (t.gc.metrics.NumBuckets < t.gc.thresholds.NumBuckets) &&
+		(t.gc.metrics.NumCalls < t.gc.thresholds.NumCalls) {
```

Comment:
Umm, these counters will need to be atomic or synchronized. If not we can have multiple GCs running at the same time or other race conditions.

Reply:
Implemented using `atomic.Uint`s in 56d29b167.